### PR TITLE
feat(new sink): New `nats` sink

### LINKED ISSUE #1399: New `nats` sink
NATS is an open-source messaging system in CNCF. I'd like to create a sink that flushes logs events to a NATS system.

### PR DESCRIPTION
**Pull request by erhlee-bird**
Introduces a new NATS log sink.
Closes #1399.

### CODE REVIEW

**Comment by erhlee-bird** [comment]
To begin with, the code is basically just a straight copy of the console sink.
I've been able to compile and confirm with a basic setup that messages are being sent to NATS.

I copied some of the structure from various PRs I found in the repo, but definitely am missing a ton of things like documentation and metadata.

**Comment by jamtur01** [comment]
Hi @erhlee-bird! Thanks for contributing. Anything we can help with here to get this out of draft?

**Comment by erhlee-bird** [comment]
Hey @jamtur01, I just started looking at this again today.

I've verified that I have a steel-thread running via hand-jammed integration tests.

I believe the next steps are

- populate .meta/sinks/nats.toml.erb with requisite data
- add in unit and integration tests
- rebase to latest
- run through the staging ci with bors?

Am I missing anything there?

**Comment by jamtur01** [comment]
We just replaced the `.meta` stuff with Cue-based docs - have a look at the docs in `docs/` for examples and we're not using Bors right now due to some bugs but otherwise, that sounds great!

**Comment by erhlee-bird** [comment]
@jamtur01, I think current is ready for review!

Thanks for the poke

**Comment by jamtur01** [comment]
@erhlee-bird Looks like you just need to run `cue fmt` on the docs:

```
cue fmt docs/**/*.cue
```

Will work.

File: Cargo.toml
Comment (praveenperera) [inline Cargo.toml:167]: Nats 0.8 was released Sept 15th

Comment (erhlee-bird) [reply Cargo.toml]: thanks, I bumped the version and ran `make check-all` before pushing

**Comment by erhlee-bird** [comment]
I'm not sure why the cue docs check is failing.
Any pointers there? I don't have any errors locally running `make check-docs`.

File: src/sinks/nats.rs
Comment (jszwedko) [inline src/sinks/nats.rs:59]: I think we should make this field templatable to let different messages flow to different topics. You can see an example of this in the `kafka` sink for the `topic` field:

https://github.com/timberio/vector/blob/2ee05ab058e4d19cb3b4fa7e0adfca92389c610a/src/sinks/kafka.rs#L135

File: src/sinks/nats.rs
Comment (jszwedko) [inline src/sinks/nats.rs:88]: I think we should split up these internal events into:

* `EventProcessed`: simply indicates an event was handled. This event is common accross our components. It should emit the counter, as you are, and also log at the trace level to be consistent.
* `NatsEventSendSucces`: indicates the event was successfully published. I like that you included the bytes sent.
* `NatsEventSendFail`: indicates an event failed to be sent. This should log the error (rate limited)

File: src/sinks/nats.rs
Comment (jszwedko) [inline src/sinks/nats.rs:38]: Could we have this healthcheck just try to connect, similar to the socket sinks?

File: src/sinks/nats.rs
Comment (jszwedko) [inline src/sinks/nats.rs:115]: This is the only subscriber, no? `vector` would be the publisher here.

I realize `nats` is "at most once", but if we start the subscriber first, we should receive all messages, no? It's just if the subscriber is offline that messages would be dropped.

File: src/sinks/nats.rs
Comment (jszwedko) [inline src/sinks/nats.rs:137]: Nit: Could we just create a standalone `nats` subscriber? It feels a little strange to create the sink just for the side-effect of the consumer. You could move the client creation into a function if you wanted to share it.

File: src/sinks/nats.rs
Comment (jszwedko) [inline src/sinks/nats.rs:87]: I'll note this in the review description, but we should avoid dropping messages that cannot be published. Instead we should continuously retry them, acking (via the sink context) once they were published. This'll give users higher delivery guarantees.

**Review by jszwedko — CHANGES_REQUESTED** [review]
This is a great, thanks @erhlee-bird . I left some inline comments to address, but it is pretty close.

In addition to the in-line comments, I'd like to see us handle failed publishes differently by:

* Using the `acker()` from the `SinkContext` (available in `build`) to only ack messages that were successfully published
* Retry failed publishes indefinitely
* Set `reconnect_buffer_size` on the nats client to 0 bytes so that the client doesn't buffer internally (to avoid message loss)

You can see the `kafka` sink for an example of the acking (open https://github.com/timberio/vector/blob/master/src/sinks/kafka.rs and search for "ack"). This will allow messages to sit in an intermediate buffer in vector until they can successfully be published to `nats`.

Let me know if that isn't clear.

Just noting that we should probably follow this work up with:
* TLS support
* Credentials support (user/pass, token, credentials file, nkey)

We can open separate issues for those once this is merged.

Thanks again!

**Comment by jszwedko** [comment]
I neglected that we aren't supporting `encoding` configuration here either. It would be nice to do that for consistency before this is merged. Again, the `kafka` sink (or the `console` sink) would be good references here.

File: docs/reference/components/sinks/nats.cue
Comment (erhlee-bird) [reply docs/reference/components/sinks/nats.cue:42]: Yeah I saw the endpoint for other examples like Pulsar that looked similar.

I set it to `url` just to try and mimic the verbiage used by the nats.go library and the nats.io docs, but totally understand if it would make more sense to stay consistent within Vector context.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs:79]: Awesome, I was wondering if there was a good way to express that covering just logs would be exhaustive for this sink.

Thank you

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs]: Code is now using the async API.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs]: Added a separate `connect` function to the SinkConfig.

Added the name config item.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs]: As best I could, I copied what Kafka had as far as the `Template` stuff goes

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs]: Added the events, but please double check me on if I ordered their emissions correctly.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs]: Yeah, it's the only subscriber with Vector being the publisher like you said. I think I had meant to say "separate."

We should, I was just overly conservatively estimating conditions that might make the integration test flaky.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs]: I think this is done. I set up the acker and acked on successful processing.
I didn't do anything in the ways of managing a message buffer myself. Is that something the acker handles?

**Comment by erhlee-bird** [comment]
@jszwedko I think I addressed all of your comments and am ready for another review.

File: src/sinks/nats.rs
Comment (jszwedko) [reply src/sinks/nats.rs]: :+1: the acking looks good. The message buffer should be handled automatically with that.

**Review by jszwedko — CHANGES_REQUESTED** [review]
This is looking good, thanks for making all of the changes!

I left a few more in-line comments to address, but this seems to be almost there.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs:131]: This is just my inexperience with Rust shining through 😄 

`nats::Options` doesn't implement the `Clone` or `Copy` traits and I couldn't figure out how to satisfy the compiler in this instance..

Comment (jszwedko) [reply src/sinks/nats.rs:131]: :smile: no worries; I'm still learning too. I think it's fine to just do `let nc = self.options.clone()` here instead. I think that'll satisfy that error. If not, could you provide the full error you are seeing?

Comment (erhlee-bird) [reply src/sinks/nats.rs:131]:
```
error[E0599]: no method named `clone` found for struct `nats::Options` in the current scope
   --> src/sinks/nats.rs:131:31
    |
131 |         let nc = self.options.clone()
    |                               ^^^^^ method not found in `nats::Options`
```

```
error[E0507]: cannot move out of `_self.options` which is behind a mutable reference
   --> src/sinks/nats.rs:131:19
    |
131 |         let nc = &self.options
    |                   ^^^^^^^^^^^^ move occurs because `_self.options` has type `nats::Options`, which does not implement the `Copy` trait
```

Comment (erhlee-bird) [reply src/sinks/nats.rs:131]:
```
error[E0507]: cannot move out of `_self.options` which is behind a mutable reference
   --> src/sinks/nats.rs:131:18
    |
131 |         let nc = self.options
    |                  ^^^^^^^^^^^^ move occurs because `_self.options` has type `nats::Options`, which does not implement the `Copy` trait
```

these were a few of the different techniques I tried

Comment (praveenperera) [reply src/sinks/nats.rs:131]: @erhlee-bird

It would be easier if here:

```
pub struct NatsSink {
...
    options: nats::Options,
...
}
```

You kept the options in your own type. One where you can clone and do whatever you want. Then when you finally create the NATS connection you can do a 

`impl From<MyNatsOption> for nats::Options`

Comment (praveenperera) [reply src/sinks/nats.rs:131]: @erhlee-bird

Here is what I meant: https://github.com/praveenperera/vector/commit/2f8fcd0aceb7d4e3417c0aea8f2b4eb965b815e4

Comment (erhlee-bird) [reply src/sinks/nats.rs]: Thanks @praveenperera, that clears things up a lot.
I need to read about the From trait, but I get the gist of things

**Review by jszwedko — COMMENTED** [review]
Thanks for the updates @erhlee-bird !

Would you mind merging in `master`? If you give edit permissions to the PR, I can do that and fix the docs (if you want). https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

The contributor DCO check is also failing which you can fix by rebasing and signing off the commits currently missing sign-off.

The integration test is currently failing for me:

```
thread 'sinks::nats::integration_tests::nats_happy' panicked at 'assertion failed: `(left == right)`
  left: `992`,
 right: `1000`', src/sinks/nats.rs:282:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test sinks::nats::integration_tests::nats_happy ... FAILED

failures:

failures:
    sinks::nats::integration_tests::nats_happy
```

It seems like this could be an issue with the sink not flushing fully before the test runs, but it seems like it should be... I can dig into this a bit more, but figured maybe you'd have more of an idea given my lack of familiarity with NATS.

File: Cargo.toml
Comment (praveenperera) [inline Cargo.toml:167]: New nats version released (0.8.2), import bug fix: 

https://github.com/nats-io/nats.rs/pull/119

**Comment by jamtur01** [comment]
@jszwedko Could you take a look at this next week please?

@erhlee-bird There's some minor merge conflicts but almost there!

File: Cargo.toml
Comment (erhlee-bird) [reply Cargo.toml]: Updated to 0.8.3 now

**Comment by erhlee-bird** [comment]
```
Commit sha: 2f8fcd0, Author: Praveen Perera, Committer: Praveen Perera; The sign-off is missing.
Commit sha: 9aab169, Author: Praveen Perera, Committer: Praveen Perera; The sign-off is missing.
```

checked the DCO bot and think I got the remaining commits that I hadn't signed off properly
@praveenperera, could you sign off these 2 commits?

**Comment by erhlee-bird** [comment]
@jszwedko, looks like the box is ticked for maintainers to freely edit 
I resolved the merge conflicts and updated to support the missing `generate_config` behaviour.

File: src/sinks/nats.rs
Comment (jszwedko) [reply src/sinks/nats.rs]: Events look good :+1:

File: src/sinks/nats.rs
Comment (jszwedko) [reply src/sinks/nats.rs]: This test seems to be passing for me now (and I note that it verifies all messages were sent and received).

**Review by jszwedko — COMMENTED** [review]
A few last comments, but otherwise this looks good! Thanks for all of your work @erhlee-bird 

And yes, @praveenperera , if you could sign-off the two commits, that would fix the DCO check.

**Comment by jszwedko** [comment]
I would like to get one more review on this just given it is a new component. @JeanMertz do you mind taking a look?

**Comment by praveenperera** [comment]
> A few last comments, but otherwise this looks good! Thanks for all of your work @erhlee-bird
> 
> And yes, @praveenperera , if you could sign-off the two commits, that would fix the DCO check.

Hey i’m not sure how to sign off on old commits. Since there have been new commits by @erhlee-bird after mine. I could rebase but don’t think I have permission to write to this branch.

File: src/sinks/nats.rs
Comment (erhlee-bird) [reply src/sinks/nats.rs:284]: That change sounds good to me.
I copied that failures pattern from the kafka test.

Monkey see, monkey do :)

**Comment by jszwedko** [comment]
@praveenperera :+1: it looks like you have access now at least? I think you can:

```
$ git rebase -i HEAD~20
# mark 2f8fcd0 and 9aab169 for edit
# when the rebase pauses, `git commit --amend --signoff && git rebase --continue`
$ git push -f
```

**Review by jszwedko — APPROVED** [review]
I pushed one more commit adding the integration tests to CI and an example of templating for the `subject` config.

Otherwise, this looks good to me! Thanks for pushing this over the line @erhlee-bird and  @praveenperera !

**Review by lukesteensen — APPROVED** [review]
Looks good to me!

**Comment by jamtur01** [comment]
@erhlee-bird Thanks! Awesome contribution.

image

**Pull request merged**
